### ISSUE TITLE
Allow configuring Docker compose to automatically start with server

### ISSUE THREAD

**Issue by jmsche** [symfony-cli/symfony-cli#240]
Hi,

I think it would be nice to have an option (in .symfony.local.yaml?) to specify that it should also run Docker compose automatically when running the server.

Maybe also allowing to define the Docker compose files to use?
Like this:

```yaml
docker_compose:
    - docker-compose.yaml
    - docker-compose.override.yaml
```

So when running `server:start` it would run `docker compose up`? And if daemon is enabled with adding the `-d` option? (so when running `server:stop` it should also run `docker compose stop` if daemon was enabled)

**Comment by mwansinck** [comment]
This can already be achieved, we currently use this in some of our projects;

```yaml
# .symfony.local.yaml
workers:
  # built-in command that builds and watches front-end assets
  # yarn_encore_watch:
  #     cmd: ['yarn', 'encore', 'dev', '--watch']
  yarn_encore_watch: ~

  docker_compose_up:
    cmd: ['docker-compose', 'up']
```

when `symfony server:start` is executed, `docker-compose up` is started as worker, when the proces is stopped, it stops all the workers, including `docker-compose up`. This also works when `symfony server:start` is running as daemon with `-d`

**Comment by tucksaun** [comment]
@mwansinck that was indeed the start of the reflexion during a chat we had with @jmsche

but I was concerned by the fact that it could cause issues with some workers being started too early (before docker compose up) and as such they could not start properly making server:start fail.
Did you experience anything similar?

**Comment by mwansinck** [comment]
> @mwansinck that was indeed the start of the reflexion during a chat we had with @jmsche
> 
> but I was concerned by the fact that it could cause issues with some workers being started too early (before docker compose up) and as such they could not start properly making server:start fail. Did you experience anything similar?

@tucksaun We do not yet have experience with that situation, but currently we are transforming our "old" docker development stacks to Symfony cli + docker, so I expected to have some experience within a few days :-)

Maybe if this will be an issue, a delay start will solve this; something like `sleep 5 && symfony console messenger:consume async`

**Comment by mwansinck** [comment]
@tucksaun I have tested the combination of messenger:consume and docker compose up, and the messenger:consume is failing due to starting to early. For now I have created the following workaround;

#bin/messenger_consume.sh
```bash
#!/bin/bash

# Wait to rabbitmq and other containers to start
sleep 15

symfony console messenger:consume async
```

# .symfony.local.yaml
```yaml
workers:
  # built-in command that builds and watches front-end assets
  # yarn_encore_watch:
  #     cmd: ['yarn', 'encore', 'dev', '--watch']
  yarn_encore_watch: ~

  docker_compose_up:
    cmd: ['docker-compose', 'up']

  messenger_consume_async:
    cmd: [ "bin/messenger_consume.sh" ]
    watch: [ "config", "src", "templates", "vendor" ]
````

The solution could be improved, possible by using the Docker API to check if the required / all container(s) are running.

**Comment by tucksaun** [comment]
thank you @mwansinck for taking the time to confirm my thoughts.

checking the required containers looks a bit complicated to me as it implies to reimplement some Docker Compose logic and my guts tells me it will be hard to maintain as we could discover more and more tons of edge cases.

It could be implemented more easily by adding an option telling the CLI to start docker compose **before** starting workers and/or the web server.
Maybe this could be another special worker key entry (as for `yarn_encore_watch` or `messenger_consume_async` that we can prioritize easily.

**Comment by mwansinck** [comment]
> It could be implemented more easily by adding an option telling the CLI to start docker compose **before** starting workers and/or the web server.

I really like this idea. The special worker is a good start, but the underling logic could be something like worker Y depends on worker X, so worker Y is waiting to be started when worker X is ready. There are some difficulties here;

For example, a worker system is meant to run infinite until Symfony Server is stopped. Thats okay when using `docker-compose up` (in the above example worker X) However, by running docker-compose without --detach (-d) we receive no exit code while running, so no way to know when worker Y should be started. When running docker-compose detached, Symfony server will restart the process over-and-over and wasting CPU time. 
To solve this, a worker should be told to run only once, but that isn't really the concept of a worker. 

Another problem here is that when the docker containers are running, the service in the container itself couldn't be ready yet. For example, when booting database (mysql/postgres) for the first time, they must initialise themself before accepting calls from the application. And while the database isn't ready, all async messenger process will fail. This problem is described by docker in their [docs](https://docs.docker.com/compose/startup-order/), but is out-of-scope for docker.

Unless Symfony CLI is going to support higher level integration with docker (through the docker API or by using some docker cli calls and parsing the output) to check if a specific container ready (healthy in case when healthcheck is configured), without a healthcheck, there is no pragmatic way to be sure a container is ready. MySQL server has a built-in healthcheck.

So the idea is good, implementation could be difficult at least..

**Comment by yguedidi** [comment]
Hello, what about embrassing the workaround by supporting a `delay` (or more explicitly `start_delay`) worker option?

**Comment by sarim** [comment]
A simple delay wouldn't work. For it to work flawlessly and not blow up things, symfony-cli needs to implement libcontainer/docker api, to act similarly as docker-compose script. Hacks like delay might look good on first try, but will break things.
Not entirely relevant, but I tried adding "docker-compose up", with yarn encore and messenger worker to .symfony.local.yaml. Now I found out I can't drop database, because messenger worker has active connection in database, To do that I have to symfony server:stop, that brings down the whole thing. I'm not complaining, just that all development tools process management is not as simple as it looks at first glance.

**Comment by yguedidi** [comment]
Inspiring from systemd, what about kind of "pre worker commands", see https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStartPre=

**Comment by tucksaun** [comment]
> To solve this, a worker should be told to run only once, but that isn't really the concept of a worker.

It's true but still this is still a concept we implemented because this is required for assets building system that does not have FS watching features so we have a concept that does not restart the command immediately if the exit status is 0.
This has been lost at some point but https://github.com/symfony-cli/symfony-cli/pull/308 should restore it.
The idea is to run the command once and only restart it on error or FS changes, so this could work for Docker Compose and even restart automatically on docker-compose.yml change

> Another problem here is that when the docker containers are running, the service in the container itself couldn't be ready yet.

But as we restart processes on error this should not be an issue I guess

> symfony-cli needs to implement libcontainer/docker api, to act similarly as docker-compose script

To me this is not maintainable on the long run because it means being on part feature wise with Docker Compose and maintain compatibility for such a nich-feature

**Comment by tucksaun** [comment]
> Not entirely relevant, but I tried adding "docker-compose up", with yarn encore and messenger worker to .symfony.local.yaml. Now I found out I can't drop database

I believe this is not related to yarn, nor adding docker compose to your `.symfony.local.yaml` but only related to the fact your worker does not close the DB connection. Enabling `doctrine_close_connection` middleware should fix your issue.
